pkg/cmd/server/start/bootstrap_node.go

```diff
@@ -49,123 +50,70 @@ func readOrCreatePrivateKey(path string) (*rsa.PrivateKey, error) {
-		ObjectMeta: metav1.ObjectMeta{
-			Name: fmt.Sprintf("node-bootstrapper-client-%s", safeSecretName(o.NodeArgs.NodeName)),
-		},
 		Spec: certificates.CertificateSigningRequestSpec{
 			Request: csrData,
+			Usages:  usages,
 		},
-	}
-
-	csr, err := c.Certificates().CertificateSigningRequests().Create(signingRequest)
+	})
 	if err != nil {
-		if !kerrors.IsAlreadyExists(err) {
-			return nil, err
-		}
-		glog.V(3).Infof("Bootstrap client certificate already exists at %s", signingRequest.Name)
-		csr, err = c.Certificates().CertificateSigningRequests().Get(signingRequest.Name, metav1.GetOptions{})
-		if err != nil {
-			return nil, err
-		}
-	}
-	if err := ioutil.WriteFile(clientKeyPath, privateKeyData, 0600); err != nil {
-		return nil, err
+		return nil, fmt.Errorf("cannot create certificate signing request: %v", err)
 	}
 
-	err = wait.PollImmediate(1*time.Second, timeout, func() (bool, error) {
-		if deny := hasCSRCondition(csr.Status.Conditions, certificates.CertificateDenied); deny != nil {
-			glog.V(2).Infof("Bootstrap signing rejected (%s): %s", deny.Reason, deny.Message)
-			return false, fmt.Errorf("certificate signing request rejected (%s): %s", deny.Reason, deny.Message)
-		}
-		if approved := hasCSRCondition(csr.Status.Conditions, certificates.CertificateApproved); approved != nil {
-			glog.V(2).Infof("Bootstrap client cert approved")
-			return true, nil
-		}
-		csr, err = c.Certificates().CertificateSigningRequests().Get(csr.Name, metav1.GetOptions{})
-		return false, err
+	// Make a default timeout = 3600s.
+	var defaultTimeoutSeconds int64 = 3600
+	certWatch, err := client.Watch(metav1.ListOptions{
+		Watch:          true,
+		TimeoutSeconds: &defaultTimeoutSeconds,
+		FieldSelector:  fields.OneTermEqualSelector("metadata.name", req.Name).String(),
 	})
 	if err != nil {
-		return nil, err
-	}
-
-	if err := ioutil.WriteFile(clientCertPath, csr.Status.Certificate, 0600); err != nil {
-		return nil, err
-	}
-
-	if _, err := tls.LoadX509KeyPair(clientCertPath, clientKeyPath); err != nil {
-		return nil, fmt.Errorf("bootstrap client certificate does not match private key, you may need to delete the client CSR: %v", err)
+		return nil, fmt.Errorf("cannot watch on the certificate signing request: %v", err)
 	}
+	defer certWatch.Stop()
+	ch := certWatch.ResultChan()
 
-	// write a kube config file for the node that contains the client cert
-	cfg, err := o.NodeArgs.KubeConnectionArgs.ClientConfig.RawConfig()
-	if err != nil {
-		return nil, err
-	}
-	if err := clientcmdapi.MinifyConfig(&cfg); err != nil {
-		return nil, err
-	}
-	ctx := cfg.Contexts[cfg.CurrentContext]
-	if len(ctx.AuthInfo) == 0 {
-		ctx.AuthInfo = "bootstrap"
-	}
-	cfg.AuthInfos = map[string]*clientcmdapi.AuthInfo{
-		ctx.AuthInfo: {
-			ClientCertificateData: csr.Status.Certificate,
-			ClientKeyData:         privateKeyData,
-		},
-	}
-	if err := clientcmd.WriteToFile(cfg, nodeConfigPath); err != nil {
-		return nil, err
-	}
+	for {
```

Comment:
> We can, although there's a more complex wrapper in the upstream client code

It shouldn't be much worse here though.  I'd like the consistency and it's easier for people read in general.